source/AI.cpp

```diff
@@ -2181,98 +2182,110 @@ void AI::KeepStation(Ship &ship, Command &command, const Body &target)
+				|| target.Velocity().Dot(-direction.Unit()) <= reverseSpeed);
+		slowdownDistance = approachSpeed * approachSpeed / (useReverse ?
+			ship.ReverseAcceleration() : (ship.Acceleration() + 160. / ship.TurnRate())) / 2.;
+
+		// If we're too close, run away.
+		if(direction.Length() <
+				max(minSafeDistance + max(slowdownDistance, 0.), useArtilleryAI * .75 * shortestArtillery))
 		{
-			command.SetTurn(TurnToward(ship, d));
-			if(ship.Facing().Unit().Dot(d) >= 0.)
-				command |= Command::BACK;
+			if(useReverse)
+			{
+				command.SetTurn(TurnToward(ship, direction));
+				if(ship.Facing().Unit().Dot(direction) >= 0.)
+					command |= Command::BACK;
+			}
+			else
+			{
+				command.SetTurn(TurnToward(ship, -direction));
+				if(ship.Facing().Unit().Dot(direction) <= 0.)
+					command |= Command::FORWARD;
+			}
 		}
 		else
 		{
-			command.SetTurn(TurnToward(ship, -d));
-			if(ship.Facing().Unit().Dot(d) <= 0.)
-				command |= Command::FORWARD;
+			// This isn't perfect, but it works well enough.
+			if((useArtilleryAI && (approachSpeed > 0. && weaponDistanceFromTarget < shortestArtillery * .9)) ||
+					weaponDistanceFromTarget < shortestRange * .75)
+				AimToAttack(ship, command, target);
+			else
+				MoveToAttack(ship, command, target);
 		}
-		return;
 	}
+	// Fire if we can or move closer to use all weapons.
+	else
+		if(weaponDistanceFromTarget < shortestRange * .75)
+			AimToAttack(ship, command, target);
+		else
+			MoveToAttack(ship, command, target);
+}
+
 
-	MoveToAttack(ship, command, target);
+
+void AI::AimToAttack(Ship &ship, Command &command, const Body &target)
+{
+	command.SetTurn(TurnToward(ship, TargetAim(ship, target)));
 }
 
 
 
 void AI::MoveToAttack(Ship &ship, Command &command, const Body &target)
 {
-	Point d = target.Position() - ship.Position();
+	Point direction = target.Position() - ship.Position();
 
 	// First of all, aim in the direction that will hit this target.
-	command.SetTurn(TurnToward(ship, TargetAim(ship, target)));
-
-	// Calculate this ship's "turning radius"; that is, the smallest circle it
-	// can make while at full speed.
-	double stepsInFullTurn = 360. / ship.TurnRate();
-	double circumference = stepsInFullTurn * ship.Velocity().Length();
-	double diameter = max(200., circumference / PI);
+	AimToAttack(ship, command, target);
 
+	const auto facing = ship.Facing().Unit().Dot(direction.Unit());
 	// If the ship has reverse thrusters and the target is behind it, we can
 	// use them to reach the target more quickly.
-	if(ship.Facing().Unit().Dot(d.Unit()) < -.75 && ship.Attributes().Get("reverse thrust"))
+	if(facing < -.75 && ship.Attributes().Get("reverse thrust"))
 		command |= Command::BACK;
 	// This isn't perfect, but it works well enough.
-	else if((ship.Facing().Unit().Dot(d) >= 0. && d.Length() > diameter)
-			|| (ship.Velocity().Dot(d) < 0. && ship.Facing().Unit().Dot(d.Unit()) >= .9))
+	else if((facing >= 0. &&
```

Comment:
I'll note that you've replaced `ship.Facing().Unit().Dot(d) >= 0.` with `facing >= 0.`, but `ship.Facing().Unit().Dot(d)` is not the same as `ship.Facing().Unit().Dot(d.Unit())`, which is what `facing` now is. Is the behavior the same? It looks like there are places in other functions which use `.Dot(d)` instead of `.Dot(d.Unit())`, so I am curious if the output is the same or not.

Reply:
I'll be honest I wasnt sure but it seemed like it was, when comparing to 0 at least

Reply:
I ran the numbers and they are the same for the purposes used. If `A.Unit().Dot(B) >= 0` then `A.Unit().Dot(B.Unit()) >= 0` as well (and same with `A.Dot(B) >= 0`. The values don't match, but the signs do. This is basically just asking, "Am I within 90 degrees of my target?" If the dot product is positive then yes, but if it's negative then no. If comparing against anything other than 0, though, then we'd want to use the unit of both points, as that lets us compare something like `A.Unit().Dot(B.Unit()) > sqrt(2)` to ask "Am I within 45 degrees of my target?"

Reply:
So should I use the same one?
It could be error prone if smbd was to write smth else than 0 honestly so idk
If I add it mb a comment?

Reply:
If someone tries using something other than 0 than what you have `facing` set to is the correct number to use, not what this was before.